### PR TITLE
add caching

### PR DESCRIPTION
**Pull request by Heidr**
Used ```Rails.cache.fetch``` to create 24h caching for ```stations_map ``` and 3h caching for ```sensors``` and ```value```.

### CODE REVIEW

File: app/services/gios_api.rb [hunk 1]
```diff
@@ -5,7 +5,9 @@ def initialize(session_city)
   end
 
   def stations_map
-    stations_map = HTTParty.get('http://api.gios.gov.pl/pjp-api/rest/station/findAll')
+    stations_map = Rails.cache.fetch(:stations_map_key, expires_in: 24.hours) do
+      HTTParty.get('http://api.gios.gov.pl/pjp-api/rest/station/findAll')
```
Comment (Quintasan) [inline app/services/gios_api.rb:9]: I think caching either the `#body` of the [Response object](https://www.rubydoc.info/github/jnunemaker/httparty/HTTParty/Response#body-instance_method) or caching the parsed JSON would be a better idea.

File: app/services/gios_api.rb [hunk 2]
```diff
@@ -112,7 +116,9 @@ def station_info(id, name)
 
   # get data from given sensor and return the latest one with valid value
   def value(id)
-    value_response = HTTParty.get("http://api.gios.gov.pl/pjp-api/rest/data/getData/#{id}")
+    value_response = Rails.cache.fetch("value_key_#{id}", expires_in: 3.hours) do
+      HTTParty.get("http://api.gios.gov.pl/pjp-api/rest/data/getData/#{id}")
```
Comment (Quintasan) [inline app/services/gios_api.rb:120]: I think caching either the #body of the Response object or caching the parsed JSON would be a better idea.

File: app/services/gios_api.rb [hunk 3]
```diff
@@ -5,7 +5,9 @@ def initialize(session_city)
   end
 
   def stations_map
-    stations_map = HTTParty.get('http://api.gios.gov.pl/pjp-api/rest/station/findAll')
+    stations_map = Rails.cache.fetch(:stations_map_key, expires_in: 24.hours) do
+      HTTParty.get('http://api.gios.gov.pl/pjp-api/rest/station/findAll')
+    end
     stations_map = JSON.parse(stations_map.response.body)
```
Comment (Quintasan) [inline app/services/gios_api.rb:11]: Generally speaking mutating variables is a bad practice – it makes reasoning about functions and testing harder. I would rewrite this as follows:

```ruby
def stations
  response = Rails.cache.fetch(:stations_map_key, expires_in: 24.hours) 
    HTTParty.get('http://api.gios.gov.pl/pjp-api/rest/station/findAll').response.body
  end
  JSON.parse(response)
end
```

File: app/services/gios_api.rb [hunk 4]
```diff
@@ -92,7 +94,9 @@ def readings
 
   # refactored to make #station_info shorter
   def sensors(id)
-    sensors = HTTParty.get("http://api.gios.gov.pl/pjp-api/rest/station/sensors/#{id}")
+    sensors = Rails.cache.fetch("sensors_key_#{id}", expires_in: 3.hours) do
+      HTTParty.get("http://api.gios.gov.pl/pjp-api/rest/station/sensors/#{id}")
+    end
     sensors = JSON.parse(sensors.body)
```
Comment (Quintasan) [inline app/services/gios_api.rb:100]: See my previous comment about mutating variables

File: app/services/gios_api.rb (hunk 3 above)
Comment (Heidr) [reply app/services/gios_api.rb:11]: I tried your code and was getting an error ```no implicit conversion of HTTParty::Response into String```.
I could make it work with ```JSON.parse(response).to_s```, but decided to do it like this:
```ruby
def stations_map
    response = Rails.cache.fetch(:stations_map_key, expires_in: 24.hours) do
      JSON.parse(
        HTTParty.get('http://api.gios.gov.pl/pjp-api/rest/station/findAll')
        .response.body
      )
    end
    response
  end
```

**Pull request merged**